optimum/habana/transformers/trainer.py

```diff
@@ -2593,91 +2581,41 @@ def _zero_model_grad(self, model):
                 model.zero_grad()
                 model._zero_grad_kwargs = {}
 
-    def get_num_items_in_batches(
-        self,
-        args,
-        epochs_trained,
-        num_train_epochs,
-        train_dataloader,
-        len_dataloader,
-        num_examples,
-        steps_trained_in_current_epoch,
-    ):
-        """
-        Calculate the number of items in each batch for all epochs during training.
-        """
-        steps_in_epoch = (
-            len_dataloader if len_dataloader is not None else args.max_steps * args.gradient_accumulation_steps
+    def get_batch_samples(self, epoch_iterator, num_batches):
+        batch_samples = []
+        num_items_in_batch = None
+        count_num_items_in_batch = (
+            # num_items_in_batch is passed to model forward
+            # https://github.com/huggingface/transformers/blob/v4.49.0/src/transformers/trainer.py#L3757
+            self.model_accepts_loss_kwargs
+            # num_items_in_batch is passed to compute_loss_func
+            # https://github.com/huggingface/transformers/blob/v4.49.0/src/transformers/trainer.py#L3773
+            or self.compute_loss_func is not None
+            # num_items_in_batch is also verified if (self.model_accepts_loss_kwargs or self.compute_loss_func)
+            # https://github.com/huggingface/transformers/blob/v4.49.0/src/transformers/trainer.py#L3790
         )
 
-        remainder = num_examples % args.gradient_accumulation_steps
-        if remainder == 0:
-            remainder = args.gradient_accumulation_steps
-
-        total_updates = steps_in_epoch // args.gradient_accumulation_steps + 1
-        if args.gradient_accumulation_steps == 1:
-            total_updates -= 1
-        global_step = 0
-
-        num_items_in_batches = []
-        for epoch in range(epochs_trained, num_train_epochs):
-            if epoch == epochs_trained and steps_trained_in_current_epoch > 0:
-                epoch_dataloader = skip_first_batches(train_dataloader, steps_trained_in_current_epoch)
-            else:
-                epoch_dataloader = train_dataloader
-
-            if hasattr(epoch_dataloader, "set_epoch"):
-                epoch_dataloader.set_epoch(epoch)
-
-            epoch_iterator = iter(epoch_dataloader)
+        for _ in range(num_batches):
             try:
-                first_batch = next(epoch_iterator)
+                batch = next(epoch_iterator)
```

Comment:
@IlyasMoutawwakil I think having `count_num_items_in_batch`  is very useful in some cases, and can prevent extra counting when unnecessary but separating the loops is not helping. 
```py
        for _ in range(num_batches):
            try:
                batch_samples += [next(epoch_iterator)]
            except StopIteration:
                break

        count_num_items_in_batch = count_num_items_in_batch and len(batch_samples) > 0 and "labels" in batch_samples[0]
        if count_num_items_in_batch:
```

The bottleneck is when you are accessing the batch.

Reply:
I'm not sure I understand what you mean by "separating the loops" 😅
So we keep pre-computing num batches doe all steps in advance ? but add `count_num_items_in_batch` condition to that ?

Reply:
What I mean is that you trying to count the num_items in one loop `for _ in range(num_batches):` different than up stream. transfomerser first get the batch_samples then in the second loop count the items. In my profiling counting in one loop like yours does not address the issue. The bottleneck is when you are accessing the data in batch. So you can do as up stream and only add your new flag which helps in some cases

Reply:
I reverted the whole thing to the original transformers code because imo, a step by step approach results in better runtime overall, allowing better overlap of compute and i/o. while a simulating the training loop and computing everything beforehand has more chance of a bottleneck, because you're iterating on the dataloader too fast.

Reply:
> The bottleneck is when you are accessing the data in batch. So you can do as up stream and only add your new flag which helps in some cases

Ah yes, yeah I addressed that in https://github.com/huggingface/optimum-habana/pull/1869#discussion_r2005290704, it's left over from my attempt to reduce time taken in the function, I can remove that.